lib/cudnn/descriptors.jl

```diff
@@ -0,0 +1,220 @@
+using Base: @__doc__
+
+
+"""
+    @cudnnDescriptor(XXX, setter=cudnnSetXXXDescriptor)
+
+Defines a new type `cudnnXXXDescriptor` with a single field `ptr::cudnnXXXDescriptor_t` and
+its constructor. The second optional argument is the function that sets the descriptor
+fields and defaults to `cudnnSetXXXDescriptor`. The constructor is memoized, i.e. when
+called with the same arguments it returns the same object rather than creating a new one.
+
+The arguments of the constructor and thus the keys to the memoization cache depend on the
+setter: If the setter has arguments `cudnnSetXXXDescriptor(ptr::cudnnXXXDescriptor_t,
+args...)`, then the constructor has `cudnnXXXDescriptor(args...)`. The user can control
+these arguments by defining a custom setter.
+"""
+macro cudnnDescriptor(x, set = Symbol("cudnnSet$(x)Descriptor"))
+    sname = Symbol("cudnn$(x)Descriptor")
+    tname = Symbol("cudnn$(x)Descriptor_t")
+    cache = Symbol("cudnn$(x)DescriptorCache")
+    create = Symbol("cudnnCreate$(x)Descriptor")
+    destroy = Symbol("cudnnDestroy$(x)Descriptor")
+    return quote
+        @__doc__ mutable struct $sname                      # needs to be mutable for finalizer
```

Comment:
why does the finaliser need it to be mutable here? It may be that we wouldn't stack allocate it, but maybe that's not a big deal?

Reply:
You can't have finalizers on non-mutable structs.